service/dap/server.go

```diff
@@ -1793,10 +1842,119 @@ func (s *Server) onReverseContinueRequest(request *dap.ReverseContinueRequest) {
+		return
+	}
+	// We need to translate the arg.Name to its evaluateName if the name
+	// refers to a field or element of a variable.
+	// https://github.com/microsoft/vscode/issues/120774
+	evaluateName, err := s.computeEvaluateName(v, arg.Name)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+		return
+	}
+
+	// By running EvalVariableInScope, we get the type info of the variable
+	// that can be accessed with the evaluateName, and ensure the variable we are
+	// trying to update is valid and accessible from the top most frame & the
+	// current goroutine.
+	goid, frame := -1, 0
+	evaluated, err := s.debugger.EvalVariableInScope(goid, frame, 0, evaluateName, DefaultLoadConfig)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to lookup variable", err.Error())
+		return
+	}
+
+	useFnCall := false
+	switch evaluated.Kind {
+	case reflect.String:
+		useFnCall = true
+	default:
+		// TODO(hyangah): it's possible to set a non-string variable using (`call i = fn()`)
+		// and we don't support it through the Set Variable request yet.
+		// If we want to support it for non-string types, we need to parse arg.Value.
+	}
+
+	if useFnCall {
+		// TODO(hyangah): function call injection currentlly allows to assign return values of
+		// a function call to variables. So, curious users would find set variable
+		// on string would accept expression like `fn()`.
+		if state, retVals, err := s.doCall(goid, frame, fmt.Sprintf("%v=%v", evaluateName, arg.Value)); err != nil {
+			s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+			return
+		} else if retVals != nil {
+			// The assignment expression isn't supposed to return values, but we got them.
+			// That indicates something went wrong (e.g. panic).
+			// TODO: isn't it simpler to do this in s.doCall?
+			s.resetHandlesForStoppedEvent()
+			s.sendStoppedEvent(state)
+
+			var r []string
+			for _, v := range retVals {
+				r = append(r, s.convertVariableToString(v))
+			}
+			msg := "interrupted"
+			if len(r) > 0 {
+				msg = "interrupted:" + strings.Join(r, ", ")
+			}
+
+			s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", msg)
+			return
+		}
+	} else {
+		if err := s.debugger.SetVariableInScope(goid, frame, 0, evaluateName, arg.Value); err != nil {
+			s.sendErrorResponse(request.Request, UnableToSetVariable, "Unable to set variable", err.Error())
+			return
+		}
+	}
+	// * Note on inconsistent state after set variable:
+	//
+	// The variable handles may be in inconsistent state - for example,
+	// let's assume there are two aliased variables pointing to the same
+	// memory and both are already loaded and cached in the variable handle.
+	// If set variable updates the memory through one of them,
```

Comment:
This is underselling it, the stack could have moved and setting a stale variable could corrupt the stack of a different goroutine.

Reply:
Agree. Updated the comment  (and a note in doCall) to include the discussion. PTAL.